### ISSUE TITLE
Add warning when ptr-to-int cast would truncate

### ISSUE THREAD

**Issue by jethrogb** [rust-lang/rust#31012]
There should be a lint that warns if you cast a pointer to any integer smaller than usize. For example to prevent cases like this:

```rust
println!("{}",std::mem::size_of::<std::fs::File> as u32);
```

**Comment by Amanieu** [comment]
In this example, a warning should only be issued when compiling to a 64-bit target, since casting a pointer to `u32` is perfectly valid on 32-bit targets. Maybe something like "Warning: cast to u32 may truncate pointer value". If truncation is really intended then the warning can be avoided by adding an intermediate `usize` cast.

**Comment by jethrogb** [comment]
There could also be a default-allow lint that complains about any pointer cast not to usize.

**Comment by nagisa** [comment]
Should be reported to [clippy](https://github.com/Manishearth/rust-clippy), I think.

**Comment by steveklabnik** [comment]
Since new lints have a big impact on users of rustc, the policy is that they should go through the RFC process like other user-facing changes. As such, I'm going to give this one a close, but if anyone comes across this ticket and wants this lint, consider adding it to clippy and/or writing up an RFC. Thanks!
